### ISSUE TITLE
Lock down TLS for all the Lando things

### ISSUE THREAD

**Issue by dustinleblanc** [lando/lando#446]
Feature/Enhancement Request
===========================
As a Lando User who regularly forces HTTPS/SSL/TLS on all applications,
I want Lando to automatically add trusted certificates to my host machine's keychain,
So that I do not get unexpected results when starting the application

This request is based on prior experience with Laravel Valet. Valet is Mac specific, which makes the problem easier for them to solve (one platform to interact with). See https://github.com/laravel/valet/blob/master/cli/Valet/Site.php#L250

A similar feature could make Lando usage a smoother experience for users that have bought into full SSL all the time. 

I intend to attempt to add this feature in PR and this issue is a placeholder until I can wrap that work up.

**Comment by pirog** [comment]
@dustinleblanc here are some notes from the slack channel in case you are interested in exploring an LE based approach

ilari [11:41 AM] 
@pirog well actually LE works locally but you need an actual domain for that because you need to use DNS based validation


pirog [11:41 AM] 
oh interesting @ilari


[11:41] 
cause we def own the `lndo.site` name


ilari [11:42 AM] 
Then you can use the DNS based validation


pirog [11:42 AM] 
i *think* i tried this before but ran into some issue


[11:43] 
which i don’t exactly remember but *i’m guessing* its because `*.lndo.site` points to `127.0.0.1` and LE was not a fan of that


ilari [11:43 AM] 
https://certbot.eff.org/docs/using.html#pre-and-post-validation-hooks
certbot.eff.org
User Guide — Certbot 0.19.0.dev0 documentation
Automatically enable HTTPS on your website with EFF's Certbot, deploying Let's Encrypt certificates. (19kB)


[11:43] 
Then there is this part: Example usage for DNS-01


[11:45] 
But there is still a but with this configuration. LE does not want to renew certs if the cert is too new. So if two users use the same domain I think it won't work :I


[11:45] 
Like if we both would have pirog.lndo.site on our locals and you would get the cert first I think that I would not get the cert. I'm not 100% sure does it work this way


[11:45] 
Haven't tested it out


[11:47] 
But have to go now. It's getting late here in Finland :slightly_smiling_face:

**Comment by davisshaver** [comment]
+1 this would be nice!

**Comment by nullvariable** [comment]
I don't think Let's Encrypt would work since dozens of people could be using a domain like `wordpress.lndo.site` at the same time. Unless you did user accounts for people and something like `site.user.lndo.site`

More likely installing a self signed certificate would work better.

**Comment by dustinleblanc** [comment]
That is exactly the route I am taking. Its a tough process, but I am
probably past the half-way point of implementation for OSX at least.
Windows and Linux have a significantly more complicated process for
trusting the certs, we'll get there though.

**Comment by hanoii** [comment]
I got here thinking about. While an out of the box solution, I think a middle local solution could also work.

The way I work around this locally is by generating a self-signed wildcard certificate with SAN support, with something like this: https://gist.github.com/hanoii/e9b4d4e70d1fa44964300e2cfd9a57c4

So I wonder if, besides getting into something this xplatform, at least have better certs on the images, with a proper wildcard SAN (for chrome) and all.

I think trusting a certificate can be simply left for the developer to handle. ON OSX is not harder that double clicking on the crt file. I think this can be simply be documented.

**Comment by dustinleblanc** [comment]
@hanoii what exactly doesn't work with the existing certs other than host trusting? I've already solved the installing a root certificate part and trusting that root cert on OSX. My next task in my implementation is to get the certs generated by the services to be signed using the trusted root CA cert. Once that's done, full trust should be good to go on OSX I believe. All that's left after that is to figure the Win/Linux logic for trusting.

The original issue report here is surrounding users who use SSL in all environments and force it within the project code, and are caught off guard when their local certs throw an error, so the whole of the issue is resolving that problem for people

**Comment by hanoii** [comment]
@dustinleblanc Maybe I assumed something wrong, I am just starting out with lando, but..

I have a drupal 8 site I am testing this out. I added the certificate that comes out of 

https://sime.lndo.site:444/

to my keychain and trusted it, and I still see the following:

<img width="840" alt="screen shot 2017-11-21 at 15 39 48" src="https://user-images.githubusercontent.com/677879/33090331-565ebd8a-ced2-11e7-8172-462a7f281825.png">

I hadn't yet tested this when I wrote my previous comment but I kind of assumed this was going to happen by looking at the certificate.

See https://www.thesslstore.com/blog/security-changes-in-chrome-58/ and related links. Basically the certificate has to have alternative names for chrome not to give the warning. I also found the "proxy" common name of the certificate, as odd.

I am using master out of yesterday, is this something new?

**Comment by dustinleblanc** [comment]
Nothing new as I haven't submitted a PR yet 😄 

Yeah, there are multiple issues with trusting the cert. I am currently working on the code to generate the service certs themselves based on a trusted root. During that process, I'll also work on addressing the other cert validation issues. My plan is for the common name to be `*.lndo.site` for the root cert, and to use the value of the user's proxy domain/ actual app domain for the common name of service certificates.

The certs are curerently being added by shell scripts within each service implementation though, so that might be a touch tricky to untangle as I work towards trying to use an unified approach with a simple js module to handle all cert generation.

@pirog, Any potential issues going that direction? I know you mentioned that Apache needs some goofy things to work, though if we are using a proxy container, do we ever need to care what Apache does inside the appserver?

**Comment by dustinleblanc** [comment]
reluctantly putting this on ice

**Comment by nullvariable** [comment]
TLDR;
[mitmproxy ](https://github.com/mitmproxy/mitmproxy/) creates their own CA which they have automated most of the install steps for, that generates certificates for the proxy on the fly, so that it can manipulate SSL protected sites without you having to ignore the certificate warning every time.

https://docs.mitmproxy.org/stable/concepts-certificates/

I'm not entirely sure what's involved in this setup, but the setup that mitmproxy uses is very slick and worked like a charm on my Windows machine. (still need to test for my Linux box). I think we could emulate this setup, and arguably this would be less complicated since we wouldn't have to sniff upstream to fill in cert details like they do.

To get this working on my Windows machine, I just visited a custom page on the proxy, and was greeted with the following screen:

![chrome_2018-03-12_08-07-47](https://user-images.githubusercontent.com/171515/37282933-7e98e6c6-25cc-11e8-8c7e-778dd9527c4c.png)

They auto downloaded the P12 file to my machine, I followed the instructions outlined, and the proxy now works seamlessly with HTTPS sites letting me see requests, intercept them, replay them etc.

In fact, if I tell the proxy to ignore certificate issues, it runs traffic from `lndo.site` urls with a nice green secure lock in Chrome.

The meat of mitmproxy's certificate stuff seems to be here: https://github.com/mitmproxy/mitmproxy/blob/236d88165499a3927caf193b884ac5e3bff16e67/mitmproxy/certs.py

Would this sort of setup work for Lando?

**Comment by dustinleblanc** [comment]
@nullvariable from a quick scan, this looks like it might have some magic for cert trusting on Windows, but everywhere else it's the same issue, we just need to be able to automate cert trusting in all environments. There is also an existential question about how Lando handles things that require sudo that @pirog and I have discussed.

Perhaps what you are suggesting is that we provide a web UI to instruct the user how to install the root CA?

**Comment by nullvariable** [comment]
hmmm, yeah, I know it's a sticky topic and it's at the edge of my personal knowledge. An onboarding page, or docs page that made this process easier (I've never yet figured out how to get certs trusted for lndo.site) could be enough. What grabbed me about this was how simple the steps for Windows seemed to be, if the process is similar or can be presented better, I think this would be sufficient.

**Comment by dustinleblanc** [comment]
Yeah the expected process for what I started building would be you'd install Lando, it'd ask for Sudo (as it does) and it will just install the root cert, all future certs would chain off it.

That really just leaves the instance of already having Lando installed and what you do there. I intended to just expose this as another command that you could run to generate them.

**Comment by pirog** [comment]
@dustinleblanc I'm going to bump the priority on this because i think its going to be pretty important for the PLD experience. I'm also going to bring #867 into this because i think its going to require some coordination and i want all the info on this initiative to be in a single ticket. Here are the high level things that we need to do to get TLS dialed "out of the box" for all the things

### 1. Cross container TLS

Independent of anything else on this list we need all our containers to share a single and trusted CA that can sign all internal certs. This should allow seamless cross-container TLS which will eliminate annoying user issues like "why doesnt my curl request work (self-signed cert exception)" and allow us to remove not-that-awesome and sometimes super verbose cert whitelisting (a la when we set up the solr service on the pantheon recipe). 

Added benefit is we can make sure this root cert exists in `~/.lando` so it can be manually whitelisted on the host machine, thus avoiding browser warnings as well.

### 2. Browser whitelisting during install

During Lando install we have `elevated/sudo` access and it would be awesome to generate the root CA at that time, place it into the proper location and then whitelist it on the host. This same thing should happen in "derivative" lando based products.

### 3. Implementation details

1. 1 Should happen regardless of whether 2 happens or not. Eg when we go to generate a cert we should check if the root CA exists and if it doesnt, we should create one. The reason for this is that all Lando users might not install lando via the package or they may be using a derivative lando product that uses a different installation mechanism. In these scenarios we can at least get cross-container internal TLS rolling and provide a root cert that is "on the host" and the user (with the help of some docs) whitelist themselves. 
2. We need to update our current add cert scripts to generate from the root CA
3. We need to remove any lingering logic that works around limitations in our currnet model (eg whitelisting the pantheon solr container)
4. For number 2 we need to find the smallest alpine image available and make sure we generate the root cert using that so we arent messing around with cross-platform differences. We then need to place the generated cert in a location on the host via clever docker voluming so that 1 will not try to generate a root cert because it will just use the one generated in the install. This whole thing should just be added to our current per-os postinstall scripts
5. After the cert is generated we want to run the OS-specific "whitelist this cert command"

I think if we do the above we should have a pretty good "out of the box" TLS experience both on our internal docker networks and from browser requests

**Comment by pirog** [comment]
@dustinleblanc i've got part 1 of this merged into master!

**Comment by OriginalEXE** [comment]
This is really cool, can't wait for the part 2, thanks guys!
